File: boj/DynamicProgramming/BOJ_11055.py
# 백준 11055 가장 큰 증가 부분 수열
# 21.02.22

import sys
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
sys.setrecursionlimit(10**6)

n = int(input())
arr = list(map(int, sys.stdin.readline().split()))
d = [0] * n
d[0] = arr[0]

ans = d[0]
for i in range(1, n):
    d[i] = arr[i]
    for j in range(0, i):
        if arr[j] < arr[i] and d[i] < d[j] + arr[i]:
            d[i] = d[j] + arr[i]
    ans = max(ans, d[i])

print(ans)

File: boj/DynamicProgramming/BOJ_11055.swift
// 백준 11055 가장 큰 증가 부분 수열
// 21.02.24

import Foundation

let n = Int(readLine()!)!
let a = readLine()!.split(separator: " ").map { Int(String($0))! }
var d = Array(repeating: 0, count: n)

d[0] = a[0]
var ans = d[0]
for i in stride(from: 1, to: n, by: 1) {
  d[i] = a[i]
  
  for j in stride(from: 0, to: i, by: 1) {
    if a[j] < a[i] && d[i] < d[j] + a[i] { d[i]  = d[j] + a[i] }
  }
  ans = max(ans, d[i])
}

print(ans)
